### ISSUE TITLE
CreateDelegate throws exception on calling function with "out var x" (Regression in Microsoft.CodeAnalysis.CSharp.Scripting 3.3.1))

### ISSUE THREAD

**Issue by OQO** [dotnet/roslyn#39565]
**Version Used**: 
Issue can be reproduced with .net Core 2.2 or .net Core 3.0 (Not tested on older .net Core versions)
with Microsoft.CodeAnalysis.CSharp.Scripting 3.3.1 (Older versions of this package do work correctly)

**Steps to Reproduce**:
Run the following program:

```C#
using Microsoft.CodeAnalysis.CSharp.Scripting;
using Microsoft.CodeAnalysis.Scripting;
using System;

namespace CSharpScripting
{
    public class HelpResult{ }
    public class ExecutionContext
    {
        public bool TryGetFastHelp(out HelpResult helpResult)
        {
            helpResult = new HelpResult();
            return true;
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            ScriptOptions so = ScriptOptions.Default
                .AddReferences(AppDomain.CurrentDomain.GetAssemblies())
                .AddImports(new string[] {
                    "CSharpScripting"
              });

            string code = @"
            if(TryGetFastHelp(out var result)){
                Console.WriteLine(result);
            }
            return true;
            ";

            var script = CSharpScript.Create<bool>(code, globalsType: typeof(ExecutionContext), options: so);
            var d = script.CreateDelegate();
            d(new ExecutionContext()).Wait();
            Console.ReadLine();
        }
    }
}
```

**Expected Behavior**:
Runs without errors.

**Actual Behavior**:
Throws exception on line `var d = script.CreateDelegate()` with message: **only methods, properties, fields and events can take a receiver (Parameter 'symbol')**

This code works fine with versions of Microsoft.CodeAnalysis.CSharp.Scripting **prior** to 3.3.1
This has been reproduced with .net Core 2.2 and .net Core 3.0

For version 3.3.1 one can explicitly mention the type as a workaround and no exception will be thrown: replace `TryGetFastHelp(out var result)` with `TryGetFastHelp(out HelpResult result)`

**Comment by OQO** [comment]
@YairHalberstadt Could you please check? This issue seems to be related to your changes for #35822 in 053a954 

@jasonmalinowski Could you please check as well since you reviewed/merge this commit?

And here is a stack trace that I forgot to add:
'''
System.AggregateException
  HResult=0x80131500
  Message=One or more errors occurred. (only methods, properties, fields and events can take a receiver
Parameter name: symbol)
  Source=System.Threading.Tasks.Parallel
  StackTrace:
   at System.Threading.Tasks.TaskReplicator.Run[TState](ReplicatableUserAction`1 action, ParallelOptions options, Boolean stopOnFirstFailure)
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.Tasks.Parallel.ThrowSingleCancellationExceptionOrOtherException(ICollection exceptions, CancellationToken cancelToken, Exception otherException)
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.For(Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamespaceSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceModuleSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSourceDeclarationDiagnostics(SyntaxTree syntaxTree, Nullable`1 filterSpanWithinTree, Func`4 locationFilterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnostics(CompilationStage stage, Boolean includeEarlierStages, DiagnosticBag diagnostics, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileMethods(CommonPEModuleBuilder moduleBuilder, Boolean emittingPdb, Boolean emitMetadataOnly, Boolean emitTestCoverageData, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream metadataPEStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, CompilationTestData testData, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, Stream metadataPEStream, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Scripting.ScriptBuilder.Build[T](Compilation compilation, DiagnosticBag diagnostics, Boolean emitDebugInformation, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Scripting.ScriptBuilder.CreateExecutor[T](ScriptCompiler compiler, Compilation compilation, Boolean emitDebugInformation, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Scripting.Script`1.GetExecutor(CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Scripting.Script`1.CreateDelegate(CancellationToken cancellationToken)
   at CSharpScripting.Program.Main(String[] args) in C:\Users\jonat\source\repos\CSharpScripting22\CSharpScripting22\Program.cs:line 39

Inner Exception 1:
ArgumentException: only methods, properties, fields and events can take a receiver
Parameter name: symbol

'''

**Comment by YairHalberstadt** [comment]
@OQO 

I'll try and take a look if I get a chance.

How did you work out this was caused by #35822? Do you know the exact line that caused the issue? That would be very helpful.

**Comment by YairHalberstadt** [comment]
Can you get the stack trace of the inner exception?

**Comment by OQO** [comment]
Sorry about missing that, didn't notice it was missing:
'''
   at Microsoft.CodeAnalysis.CSharp.Symbols.SymbolExtensions.RequiresInstanceReceiver(Symbol symbol)
   at Microsoft.CodeAnalysis.CSharp.Binder.TryBindInteractiveReceiver(SyntaxNode syntax, Symbol currentMember, NamedTypeSymbol currentType, NamedTypeSymbol memberDeclaringType)
   at Microsoft.CodeAnalysis.CSharp.Binder.SynthesizeMethodGroupReceiver(CSharpSyntaxNode syntax, ArrayBuilder`1 members)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindIdentifier(SimpleNameSyntax node, Boolean invoked, Boolean indexed, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindMethodGroup(ExpressionSyntax node, Boolean invoked, Boolean indexed, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindInvocationExpression(InvocationExpressionSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(ExpressionSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.GlobalExpressionVariable.InferrableGlobalExpressionVariable.InferFieldType(ConsList`1 fieldsBeingBound, Binder binder)
   at Microsoft.CodeAnalysis.CSharp.Symbols.GlobalExpressionVariable.GetFieldType(ConsList`1 fieldsBeingBound)
   at Microsoft.CodeAnalysis.CSharp.Symbols.FieldSymbol.get_TypeWithAnnotations()
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceFieldSymbol.AfterAddingTypeMembersChecks(ConversionsBase conversions, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.CheckSpecialMemberErrors(DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.AfterMembersChecks(DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Symbol.ForceCompleteMemberByLocation(SourceLocation locationOpt, Symbol member, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamespaceSymbol.<>c__DisplayClass49_1.<ForceComplete>b__0(Int32 i)
   at Roslyn.Utilities.UICultureUtilities.<>c__DisplayClass6_0`1.<WithCurrentUICulture>b__0(T param)
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw(Exception source)
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica`1.ExecuteAction(Boolean& yieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
'''

**Comment by OQO** [comment]
> @OQO
> 
> I'll try and take a look if I get a chance.
> 
> How did you work out this was caused by #35822? Do you know the exact line that caused the issue? That would be very helpful.

See the stack trace, it throws the exception in the new code added. And this used to work before these changes, of course there is a chance that it is due to another unrelated change somewhere else that I have missed. Please also see the work around that explicitly mentioning the type (instead of var) resolves the issue.

**Comment by OQO** [comment]
@jinujoseph thanks for labelling. 

Please try to give bug higher priority if possible since this is a regression and not a bug in new functionality. This bug impacts code that is in production, and had previously always worked correctly. Since exceptions happen while running the code live, not during build of the system it has real impact.

**Comment by YairHalberstadt** [comment]
@OQO

I'm working on a fix right now. I don't think this should be too complicated.

You'll be able to use the fix as soon as a nuget package is released with the changes. I don't know how that works, but looking at https://www.nuget.org/packages/Microsoft.CodeAnalysis.CSharp.Scripting/ it seems like it's quite frequent.

For now, you can simply use an older version of the package.

**Comment by OQO** [comment]
@YairHalberstadt toda raba. 

We can't use another version due to some other other code having dependencies as well, needing at least 3.3.1. 

We now made a special tool that will compile all our scripts so that we can test/make fixes and not encounter issues in production. Thanks again for looking into this issue so quickly!

**Comment by YairHalberstadt** [comment]
Fix in #39584
